genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/FileSystemFetchingCacheService.java

```diff
@@ -0,0 +1,343 @@
+                throw new RuntimeException("Failed to create directory: " + dir.getAbsolutePath(), e);
+            }
+        } else if (!dir.isDirectory()) {
+            throw new RuntimeException("This location is not a directory: " + dir.getAbsolutePath());
+        }
+    }
+
+    /**
+     * Clean up all resources older than the latest version of a resource.
+     *
+     * @param resourceCacheId                         resource cache id
+     * @param lastDownloadedResourceModifiedTimestamp timestamp of last successfully downloaded resource
+     * @throws IOException
+     */
+    @VisibleForTesting
+    void cleanUpOlderResourceVersions(
+        final String resourceCacheId,
+        final long lastDownloadedResourceModifiedTimestamp
+    ) throws IOException {
+
+        //Get all versions of a resource in the cache
+        final File[] files = getCacheResourceDir(resourceCacheId).listFiles();
+
+        if (files != null) {
+            for (File file : files) {
+                long resourceLastModified = 0;
+                try {
+                    resourceLastModified = getResourceLastModified(file);
+                } catch (NumberFormatException e) {
+                    log.warn("Encountered a dir name which is not long. Ignoring dir - {}", resourceLastModified, e);
+                }
+
+                if (resourceLastModified < lastDownloadedResourceModifiedTimestamp) {
+                    cleanUpResourceVersion(file);
+                }
+            }
+        }
+    }
+
+    /**
+     * Delete a resource version directory after taking appropriate lock.
+     *
+     * @param resourceVersionDir Directory to be deleted
+     * @throws IOException
+     */
+    private void cleanUpResourceVersion(final File resourceVersionDir) throws IOException {
+
+        try (
+            final FileChannel channel = getLockFileChannel(resourceVersionDir);
+            final Lock lock = lockingFactory.getLock(channel);
+        ) {
+            lock.lock();
+            //lock gets released right after deletion
+            FileSystemUtils.deleteRecursively(resourceVersionDir);
+        }
+    }
+
+    /* Returns a handle to the directory for a resource */
+    private File getCacheResourceDir(final String resourceCacheId) {
+        return new File(cacheDirectory, resourceCacheId);
+    }
+
+    /* Returns a handle to the directory for a resource version */
+    @VisibleForTesting
+    File getCacheResourceVersionDir(final String resourceCacheId, final long lastModifiedTimestamp) {
+        return new File(getCacheResourceDir(resourceCacheId), Long.toString(lastModifiedTimestamp));
+    }
+
+    /* Returns a handle to the data file of a resource version in the cache */
+    @VisibleForTesting
+    File getCacheResourceVersionDataFile(final String resourceCacheId, final long lastModifiedTimestamp) {
+        return new File(getCacheResourceVersionDir(resourceCacheId, lastModifiedTimestamp), "data");
```

Comment:
You created constants for "data" and "data.tmp" but you don't use them in these two methods.

Reply:
Thanks for pointing. Will fix.